src/firstboot/firstboot.c

```diff
@@ -582,23 +586,57 @@ static int prompt_root_password(void) {
         return 0;
 }
 
-static int write_root_shadow(const char *path, const struct spwd *p) {
-        _cleanup_fclose_ FILE *f = NULL;
+static int write_root_shadow(const char *shadow_path, const struct spwd *p) {
+        _cleanup_fclose_ FILE *original = NULL, *shadow = NULL;
+        _cleanup_(unlink_and_freep) char *shadow_tmp = NULL;
+        struct spwd *sp = NULL;
```

Comment:
this gets quite cryptic, given that we have both p and sp now, which are of the same type. Maybe call the argument root_sp, and the other iterator_sp or so?

Reply:
also move to innermost scope where it is needed?